CI: Fixed TestPodCIDRAllocatorOverlap

dylandreimerink opened the pull request:
Added an extra channel to the Pod CIDR allocator which is triggered after each resync. The `TestPodCIDRAllocatorOverlap` test has been updated to block on this channel three times which is when it is done syncing all changes to the API server.

Fixed: #21956

```release-note
Made the `TestPodCIDRAllocatorOverlap` test more robust
```

gandro commented on operator/cmd/allocator_test.go:
```diff
@@ -128,11 +96,15 @@ func podCIDRAllocatorOverlapTestRun(t *testing.T) {
-	<-updated
+	// Now wait for two resync events, after which the API server is in sync
+	<-podCidrManager.K8sReSyncChan
+	<-podCidrManager.K8sReSyncChan
```
Why do we expect two resyncs here?

Instead of relying on those notification that seem to rely in implementation behavior, could we change this unit test to just poll the mock APIServer and wait for changes, e.g. using [`testutils.WaitUntil`](https://github.com/cilium/cilium/blob/314ca7baeff4f568ffc0bad95124e40665b1f88c/pkg/testutils/condition.go#L19)

dylandreimerink replied:
I am not 100% on which event is triggered by which code path, this turned out to work experimentally. It gets a bit muddy since [syncNode](https://github.com/cilium/cilium/blob/6d561885ba40364491122cbc5ca92d54f09b808a/pkg/ipam/allocator/podcidr/podcidr.go#L576) also triggers the k8sReSync which itself is called at a number of locations. 

> could we change this unit test to just poll the mock APIServer and wait for changes

This is what the code did that I am currently replacing. You have to have a timeout when you give up, it is hard to set such a timeout since time seems to warp in the CI due to congestion(I suspect we get way less cpu cycles per second, so we start missing timeouts that otherwise make sense). This is an attempt to have a trigger that isn't based on time.

Though I have to agree that this is not ideal in case the implementation ever changes. :sigh:. I am starting to think that we should just remove this test all together, it is costing way to many cycles than it is worth.

gandro replied:
I agree that this is incredibly hard to test.

Could we simplify it, i.e. not trying to wait for the API server to become stable or observe resyncs. Instead, just poll nodeA and nodeB until they have the expected state (or fail if we detect an overlap)? Or would that never uncover the bug fixed by your previous PR?

gandro replied:
To be a bit more precise, I envision something along the lines of:

```go
	// Wait for the "node manager synced" signal, just like we would normally.
	<-ciliumNodeManagerQueueSynced

	// Trigger the Resync after the cache sync signal
	podCidrManager.Resync(ctx, time.Time{})

	err := testutils.WaitUntil(func() bool {
		// Get node A from the mock APIServer
		nodeAInt, err := fakeClient.Tracker().Get(ciliumnodesResource, "", "node-a")
		if err != nil {
			return false
		}
		nodeA := nodeAInt.(*cilium_api_v2.CiliumNode)

		// Get node B from the mock APIServer
		nodeBInt, err := fakeClient.Tracker().Get(ciliumnodesResource, "", "node-b")
		if err != nil {
			return false
		}
		nodeB := nodeBInt.(*cilium_api_v2.CiliumNode)

		if len(nodeA.Spec.IPAM.PodCIDRs) != 1 {
			return false
		}

		if len(nodeB.Spec.IPAM.PodCIDRs) != 1 {
			return false
		}

		// The PodCIDRs should be distinct.
		if nodeA.Spec.IPAM.PodCIDRs[0] == nodeB.Spec.IPAM.PodCIDRs[0] {
			t.Fatal("Node A and Node B are assigned overlapping PodCIDRs")
		}
		
		return true
	}, 2*time.Minute)
	if err != nil {
		t.Fatalf("nodes have no pod CIDR: %s", err)
	}
```

dylandreimerink replied:
I don't know if you tested this before hand, but it works beautifully. Drop in replacement!

aanm reviewed (commented):
I'll default to Sebastian's review

dylandreimerink committed:
CI: Fixed TestPodCIDRAllocatorOverlap

Changed the test to keep polling the mock API server until both nodes
have been updated and then perform the actual assertion logic, otherwise
block 2 minutes.

Fixed: #21956
Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
